File: src/crypto/plugin/CryptoFactory.ts
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

import CryptoOperations, { CryptoSuiteMap } from './CryptoOperations';
import SubtleCryptoBrowserOperations from './SubtleCryptoBrowserOperations';
import IKeyStore from '../keyStore/IKeyStore';
import { SubtleCrypto } from 'webcrypto-core';

/**
 * Utility class to handle all CryptoSuite dependency injection
 */
export default class CryptoFactory {

  /**
   * The key encryptors
   */
  public keyEncrypters: CryptoSuiteMap;

  /**
   * The shared key encryptors
   */
  public sharedKeyEncrypters: CryptoSuiteMap;
  
  /**
   * The symmetric content encryptors
   */
  public symmetricEncrypter: CryptoSuiteMap;
  
  /**
   * The message signer
   */
  public messageSigners: CryptoSuiteMap;
  
  /**
   * The hmac operations
   */
  public messageAuthenticationCodeSigners: CryptoSuiteMap;
  
  /**
   * The digest operations
   */
  public messageDigests: CryptoSuiteMap;

  /**
   * Key store used by the CryptoFactory
   */
  public keyStore: IKeyStore;

  /**
   * Label for default algorithm
   */
  private readonly defaultAlgorithm = '*';

  /**
   * Constructs a new CryptoRegistry
   * @param keyStore used to store private jeys
   * @param crypto The suite to use for dependency injection
   */
  constructor (keyStore: IKeyStore, crypto: CryptoOperations) {
    this.keyStore = keyStore;
    this.keyEncrypters = <CryptoSuiteMap>{'*': crypto };
    this.sharedKeyEncrypters = <CryptoSuiteMap>{'*': crypto };
    this.symmetricEncrypter = <CryptoSuiteMap>{'*': crypto };
    this.messageSigners = <CryptoSuiteMap>{'*': crypto };
    this.messageAuthenticationCodeSigners = <CryptoSuiteMap>{'*': crypto };
    this.messageDigests = <CryptoSuiteMap>{'*': crypto };
  }

  /**
   * Gets the key encrypter object given the encryption algorithm's name
   * @param name The name of the algorithm
   * @returns The corresponding crypto API
   */
  public getKeyEncrypter (name: string): SubtleCrypto {
    if (this.keyEncrypters[name]) {
      return this.keyEncrypters[name].getKeyEncrypters();
    }
    return this.keyEncrypters[this.defaultAlgorithm].getKeyEncrypters();
  }

  /**
   * Gets the shared key encrypter object given the encryption algorithm's name
   * Used for DH algorithms
   * @param name The name of the algorithm
   * @returns The corresponding crypto API
   */
  getSharedKeyEncrypter (name: string): SubtleCrypto {
    if (this.sharedKeyEncrypters[name]) {
      return this.sharedKeyEncrypters[name].getSharedKeyEncrypters();
    }
    return this.sharedKeyEncrypters[this.defaultAlgorithm].getSharedKeyEncrypters();
  }

  /**
   * Gets the SymmetricEncrypter object given the symmetric encryption algorithm's name
   * @param name The name of the algorithm
   * @returns The corresponding crypto API
   */
  getSymmetricEncrypter (name: string): SubtleCrypto {
    if (this.symmetricEncrypter[name]) {
      return this.symmetricEncrypter[name].getSymmetricEncrypters();
    }
    return this.symmetricEncrypter[this.defaultAlgorithm].getSymmetricEncrypters();
  }
  
  /**
   * Gets the message signer object given the signing algorithm's name
   * @param name The name of the algorithm
   * @returns The corresponding crypto API
   */
  getMessageSigner (name: string): SubtleCrypto {
    if (this.messageSigners[name]) {
      return this.messageSigners[name].getMessageSigners();
    }
    return this.messageSigners[this.defaultAlgorithm].getMessageSigners();
  }

  /**
   * Gets the mac signer object given the signing algorithm's name
   * @param name The name of the algorithm
   * @returns The corresponding crypto API
   */
  getMessageAuthenticationCodeSigners (name: string): SubtleCrypto {
    if (this.messageAuthenticationCodeSigners[name]) {
      return this.messageAuthenticationCodeSigners[name].messageAuthenticationCodeSigners();
    }
    return this.messageAuthenticationCodeSigners[this.defaultAlgorithm].messageAuthenticationCodeSigners();
  }

  /**
   * Gets the message digest object given the digest algorithm's name
   * @param name The name of the algorithm
   * @returns The corresponding crypto API
   */
  getMessageDigest (name: string): SubtleCrypto {
    if (this.messageDigests[name]) {
      return this.messageDigests[name].getMessageDigests();
    }
    return this.messageDigests[this.defaultAlgorithm].getMessageDigests();
  }
}

File: kotlin/useragent-sdk/src/commonMain/kotlin/com/microsoft/did/sdk/crypto/plugins/CryptoFactory.kt
package com.microsoft.did.sdk.crypto.plugins

import com.microsoft.did.sdk.crypto.keyStore.IKeyStore
import com.microsoft.did.sdk.crypto.models.webCryptoApi.SubtleCrypto

/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

/**
 * Utility class to handle all CryptoSuite dependency injection
 */
class CryptoFactory(keyStore: IKeyStore, crypto: CryptoOperations) {

    /**
     * The key encryptors
     */
    var keyEncrypters: Map<String, CryptoOperations> = mapOf("*" to crypto)

    /**
     * The shared key encryptors
     */
    var sharedKeyEncrypters: Map<String, CryptoOperations> = mapOf("*" to crypto)

    /**
     * The symmetric content encryptors
     */
    var symmetricEncrypter: Map<String, CryptoOperations> = mapOf("*" to crypto)

    /**
     * The message signer
     */
    var messageSigners: Map<String, CryptoOperations> = mapOf("*" to crypto)

    /**
     * The hmac operations
     */
    var messageAuthenticationCodeSigners: Map<String, CryptoOperations> = mapOf("*" to crypto)

    /**
     * The digest operations
     */
    var messageDigests: Map<String, CryptoOperations> = mapOf("*" to crypto)

    /**
     * Key store used by the CryptoFactory
     */
    var keyStore: IKeyStore = keyStore

    /**
     * Label for default algorithm
     */
    private val defaultAlgorithm: String = "*"

    /**
     * Gets the key encrypter object given the encryption algorithm's name
     * @param name The name of the algorithm
     * @returns The corresponding crypto API
     */
    fun getKeyEncrypter (name: String): SubtleCrypto {
        if (this.keyEncrypters.containsKey(name)) {
            return (this.keyEncrypters[name] ?: error("")).getKeyEncrypters();
        }
        return (this.keyEncrypters[this.defaultAlgorithm] ?: error("Default algorithm not implemented " +
                "for key encrypters")).getKeyEncrypters();
    }

    /**
     * Gets the shared key encrypter object given the encryption algorithm's name
     * Used for DH algorithms
     * @param name The name of the algorithm
     * @returns The corresponding crypto API
     */
    fun getSharedKeyEncrypter (name: String): SubtleCrypto {
        if (this.sharedKeyEncrypters.containsKey(name)) {
            return (this.sharedKeyEncrypters[name] ?: error("")).getSharedKeyEncrypters();
        }
        return (this.sharedKeyEncrypters[this.defaultAlgorithm] ?: error("Default algorithm not implemented " +
                "for shared key encrypters")).getSharedKeyEncrypters();
    }

    /**
     * Gets the SymmetricEncrypter object given the symmetric encryption algorithm's name
     * @param name The name of the algorithm
     * @returns The corresponding crypto API
     */
    fun getSymmetricEncrypter (name: String): SubtleCrypto {
        if (this.symmetricEncrypter.containsKey(name)) {
            return (this.symmetricEncrypter[name] ?: error("")).getSymmetricEncrypters();
        }
        return (this.symmetricEncrypter[this.defaultAlgorithm] ?: error("Default algorithm not implemented for " +
                "symmetric encrypters")).getSymmetricEncrypters();
    }

    /**
     * Gets the message signer object given the signing algorithm's name
     * @param name The name of the algorithm
     * @returns The corresponding crypto API
     */
    fun getMessageSigner (name: String): SubtleCrypto {
        if (this.messageSigners.containsKey(name)) {
            return (this.messageSigners[name] ?: error("")).getMessageSigners();
        }
        return (this.messageSigners[this.defaultAlgorithm] ?: error("Default algorithm not implemented for " +
                "message signers")).getMessageSigners();
    }

    /**
     * Gets the mac signer object given the signing algorithm's name
     * @param name The name of the algorithm
     * @returns The corresponding crypto API
     */
    fun getMessageAuthenticationCodeSigners (name: String): SubtleCrypto {
        if (this.messageAuthenticationCodeSigners.containsKey(name)) {
            return (this.messageAuthenticationCodeSigners[name] ?: error("")).messageAuthenticationCodeSigners();
        }
        return (this.messageAuthenticationCodeSigners[this.defaultAlgorithm] ?: error("Default algorithm not " +
                "implemented for message authentication code signers")).messageAuthenticationCodeSigners();
    }

    /**
     * Gets the message digest object given the digest algorithm's name
     * @param name The name of the algorithm
     * @returns The corresponding crypto API
     */
    fun getMessageDigest (name: String): SubtleCrypto {
        if (this.messageDigests.containsKey(name)) {
            return this.messageDigests[name]!!.getMessageDigests();
        }
        return (this.messageDigests[this.defaultAlgorithm] ?: error("Default algorithm not implemented " +
                "for message digests")).getMessageDigests();
    }
}